src/main/java/com/zendesk/maxwell/schema/MysqlSchemaCompactor.java

```diff
@@ -0,0 +1,197 @@
+package com.zendesk.maxwell.schema;
+
+import com.zendesk.maxwell.CaseSensitivity;
+import com.zendesk.maxwell.replication.Position;
+import com.zendesk.maxwell.schema.ddl.InvalidSchemaError;
+import com.zendesk.maxwell.util.ConnectionPool;
+import com.zendesk.maxwell.util.RunLoopProcess;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.sql.Connection;
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.sql.Statement;
+
+public class MysqlSchemaCompactor extends RunLoopProcess {
+	static final Logger LOGGER = LoggerFactory.getLogger(MysqlSchemaCompactor.class);
+
+	private final ConnectionPool maxwellConnectionPool;
+	private final String clientID;
+	private final Long serverID;
+	private final CaseSensitivity sensitivity;
+	private final int maxDeltas;
+
+	public MysqlSchemaCompactor(
+			int maxDeltas,
+			ConnectionPool maxwellConnectionPool,
+			String clientID,
+			Long serverID,
+			CaseSensitivity sensitivity
+	) {
+		this.maxDeltas = maxDeltas;
+		this.maxwellConnectionPool = maxwellConnectionPool;
+		this.clientID = clientID;
+		this.serverID = serverID;
+		this.sensitivity = sensitivity;
+	}
+
+	@Override
+	protected void work() throws Exception {
+		try {
+			doWork();
+			Thread.sleep(5000);
+		} catch ( InterruptedException e ) {
+		} catch ( SQLException e ) {
+			LOGGER.error("got SQLException trying to compact", e);
+		}
+	}
+
+	private String lockName() {
+		return "maxwell_schema_compaction-" + this.serverID;
+	}
+
+	private boolean getLock(Connection cx) throws SQLException {
+		PreparedStatement s = cx.prepareStatement(
+			"SELECT GET_LOCK(?, 0)"
+		);
+		s.setString(1, this.lockName());
+		ResultSet rs = s.executeQuery();
+		return rs.next() && rs.getBoolean(1);
+	}
+
+	private void releaseLock(Connection cx) throws SQLException {
+		PreparedStatement s = cx.prepareStatement(
+			"SELECT RELEASE_LOCK(?)"
+		);
+		s.setString(1, this.lockName());
+		s.execute();
+	}
+
+	public void doWork() throws Exception {
+		try ( Connection cx = maxwellConnectionPool.getConnection() ) {
+			cx.setAutoCommit(false);
+			try {
+				if (getLock(cx)) {
+					compact(cx);
+				}
+			} finally {
+				releaseLock(cx);
+			}
+		}
+	}
+
+	private boolean shouldCompact(Connection cx) throws SQLException {
+		ResultSet rs = cx.prepareStatement(
+			"select count(*) as count from `schemas` where `server_id` = " + this.serverID
+		).executeQuery();
+		return rs.next() && rs.getInt("count") >= this.maxDeltas;
+	}
+
+	Long lastWarnedSchemaID = null;
+
+	private Long chooseCompactedSchemaBase(Connection cx) throws SQLException {
+		if ( !shouldCompact(cx) ) {
+			return null;
+		}
+
+		ResultSet rs = cx.prepareStatement(
+			"select id, binlog_file, binlog_position, gtid_set, 0 as last_heartbeat_read "
+			+ " from `schemas` where `server_id` = " + this.serverID
+			+ " order by id desc limit 1"
+		).executeQuery();
```

Comment:
Looks like this duplicates but doesn't fully replicate the logic in https://github.com/zendesk/maxwell/blob/e48b0fc957c33062b32d23c21ddfb6a26b39191d/src/main/java/com/zendesk/maxwell/schema/MysqlSavedSchema.java#L578-L588

Can we share them?

Reply:
mmm, no, one asks "most-recent-for-this-position" and one asks "most-recent-overall".